operations/secure_random/secure_random.cc

```diff
@@ -0,0 +1,247 @@
+
+  // Set output shape
+  ShapeHandle out;
+  TF_RETURN_IF_ERROR(context->MakeShapeFromShapeTensor(0, &out));
+  context->set_output(0, out);
+  return Status::OK();
+}
+
+REGISTER_OP("SecureRandom")
+    .Input("shape: T")
+    .Input("seed: Tseed")
+    .Input("minval: dtype")
+    .Input("maxval: dtype")
+    .Output("output: dtype")
+    .Attr("dtype: {int32, int64} = DT_INT32")
+    .Attr("T: {int32, int64} = DT_INT32")
+    .Attr("Tseed: {int32} = DT_INT32")
+    .SetShapeFn(SecureRandomShape);
+
+// this function allows you to skip ahead in the chacha stream so you don't
+// have to allocate more memory than you need to, used in rejection sampling and
+// will be handy for parallelizing this operation
+void randombytes_buf_deterministic_ic(void * const buf, const size_t size, uint32 count,
+                              const unsigned char seed[randombytes_SEEDBYTES])
+{
+    static const unsigned char nonce[crypto_stream_chacha20_ietf_NONCEBYTES] = {
+        'L', 'i', 'b', 's', 'o', 'd', 'i', 'u', 'm', 'D', 'R', 'G'
+    };
+
+    unsigned char * u_buf = (unsigned char *)buf;
+
+    memset(u_buf, 0, size);
+
+    crypto_stream_chacha20_ietf_xor_ic(u_buf, u_buf, (unsigned long long) size,
+                                       nonce, count, seed);
+}
+
+
+template <typename T, typename Wide>
+class Generator {
+public:
+  Tensor *output = NULL;
+  const unsigned char * seeds = NULL;
+
+  Generator(Tensor* output, const unsigned char * seeds) : output(output), seeds(seeds) {
+    auto flat = output->flat<T>();
+
+    count_ = flat.size();
+    bytes_count_ = count_ * sizeof(T);
+    buf_ = static_cast<T *>(flat.data());
+
+    elements_per_block_ = CHACHABLOCKSIZE / sizeof(T);
+    extra_block_ = static_cast<T *>(malloc(CHACHABLOCKSIZE));
+
+    block_counter_ = bytes_count_ / CHACHABLOCKSIZE + 1;
+
+    // prepare the extra block if any values get rejected in the rejection sampling
+    randombytes_buf_deterministic_ic(extra_block_, CHACHABLOCKSIZE, block_counter_, seeds);
+  }
+
+  ~ Generator() {
+    free(extra_block_);
+  }
+
+  void GenerateData(T minval, T maxval) {
+    auto flat = output->flat<T>();
+
+    randombytes_buf_deterministic(buf_, bytes_count_, seeds);
+
+    Uniform(minval, maxval - 1);
+
+    std::copy(buf_, buf_ + flat.size(), flat.data());
+  }
+
+private:
+  T *buf_ = NULL;
+  uint32 count_ = 0;
+  uint32 bytes_count_ = 0;
+
+  T * extra_block_ = NULL;
+  uint32 block_counter_ = 0;
+  uint32 elements_per_block_ = 0;
+  uint32 inner_block_index_ = 0;
+
+  // The following random uniform distribution is based on a an implementation from
+  // https://github.com/rust-random/rand/blob/3eadab75c8a5871d1be729091795a6c4e1dc19bb/src/distributions/uniform.rs#L310
+  // There is quite a bit of documentation at that link which is explains the implementation.
+  // See below for other inline docs.
+
+  // inclusive uniform!
+  void Uniform(T low, T high) {
+    typedef typename std::make_unsigned<T>::type uT;
```

Comment:
this gives the unsigned version of the signed type?

Reply:
or, how is the fact that the rust code takes four types expressed here?

Reply:
> this gives the unsigned version of the signed type?

Yes, whether it's int32 or int64.

> or, how is the fact that the rust code takes four types expressed here?

This version doesn't need the four types because in the rust code all four types were being used as an optimizations for smaller width values (int8 and int16). We're only handling int32 and int64 which those optimizations are not needed.